torch/csrc/jit/ir.cpp

```diff
@@ -461,13 +461,107 @@ void Block::cloneFrom(Block * src, std::function<Value*(Value*)> value_map) {
   }
 }
 
+bool Value::equals(const Value* other) const {
+  return *this->type_ == *other->type_;
+}
+
+// Equivalent to C++14's std::equal on two ranges.
+template <typename T>
+static bool equal(
+    const at::ArrayRef<T> a,
+    const at::ArrayRef<T> b,
+    std::function<bool(T, T)> predicate) {
+  if (a.size() != b.size()) {
+    return false;
+  }
+
+  return std::equal(a.cbegin(), a.cend(), b.cbegin(), predicate);
+}
+
+bool Block::equals(
+    const Block* other,
+    std::unordered_map<const Value*, const Value*>& selfToOther) const {
+  // Define some lambdas to perform equivalence checking.
+  const std::function<bool(const Value*, const Value*)> checkValuesMatch =
+      [&](const Value* a, const Value* b) { return selfToOther.at(a) == b; };
+
+  const std::function<bool(const Value*, const Value*)> matchValues =
+      [&](const Value* a, const Value* b) {
+        if (!a->equals(b)) {
+          return false;
+        }
+        selfToOther[a] = b;
+        return true;
+      };
+
+  const std::function<bool(const Block*, const Block*)> matchBlocks =
+      [&](const Block* a, const Block* b) { return a->equals(b, selfToOther); };
+
+  // Map graph inputs to each other
+  if (!equal(this->inputs(), other->inputs(), matchValues)) {
+    return false;
+  }
+
+  // For each node in the graph...
+  auto otherIter = other->nodes().begin();
+  for (auto selfNode : this->nodes()) {
+    if (otherIter == other->nodes().end()) {
+      // otherNodes.size() < selfNodes.size()
+      return false;
+    }
+
+    auto otherNode = *otherIter;
+    // Check the node type is the same
+    if (selfNode->kind() != otherNode->kind()) {
+      return false;
+    }
+
+    // Check that node inputs are equivalent
+    if (!equal(selfNode->inputs(), otherNode->inputs(), checkValuesMatch)) {
+      return false;
+    }
+
+    // Check blocks are equivalent
+    if (!equal(selfNode->blocks(), otherNode->blocks(), matchBlocks)) {
+      return false;
+    }
+
+    // Check attributes are equivalent
+    if (!selfNode->attributesEqual(*otherNode)) {
+      return false;
+    }
+
+    // Map outputs to each other
+    if (!equal(selfNode->outputs(), otherNode->outputs(), matchValues)) {
+      return false;
+    }
```

Comment:
Those are the things that would be useful to have as `Node::equal` 😊

Reply:
I thought it might be sort of weird to put this stuff in a `Node::equals`, because this kind of equality check doesn't make sense out side of the context of a block (since we're mostly checking the flow of values through the graph). So any `Node::equals` would need to take and modify the value->value map, which seems confusing.